File: Exavision.Seasense.Shared/Capabilities/DoubleValueCapability.cs
using Exavision.Seasense.Shared.Settings;
using Exavision.Seasense.Shared.States;

namespace Exavision.Seasense.Shared.Capabilities {
    public abstract class DoubleValueCapability : Capability<SettingDoubleValueCapability> {

        public double MinValue { get; protected set; }

        public double MaxValue { get; protected set; }

        public double Value { get; protected set; }

        public abstract DoubleValueType DoubleValueType { get; }

        public override SettingDoubleValueCapability GetSetting(SettingDoubleValueCapability setting) {
            setting.MaxValue = this.MaxValue;
            setting.MinValue = this.MinValue;
            setting.Id = this.Id;
            setting.CapabilityType = CapabilityType.DoubleValue;
            setting.DoubleValueType = this.DoubleValueType;
            setting.DisplayName = this.DoubleValueType.ToString();
            return setting;
        }

        public override void SetSetting(SettingDoubleValueCapability setting) {
            base.SetSetting(setting);
            this.MaxValue = setting.MaxValue;
            this.MinValue = setting.MinValue;
        }

        public override CapabilityState GetState() {
            DoubleValueState state = new DoubleValueState() { Id = this.Id, Value = this.Value };
            return state;
        }

        public virtual void SetValue(double value) {
            this.Value = value;
        }

    }

}

File: Exavision.Seasense/Client/src/app/materials/capabilities/double-value-capability.ts
import { Material } from "../material";
import { SettingCapability } from "../settings/setting-capability";
import { SettingDoubleValueCapability } from "../settings/setting-double-value-capability";

import { CapabilityState } from "../states/capability-state";
import { DoubleValueState } from "../states/double-value-state";
import { Capability } from "./capability";
import { CapabilityType } from "./capability-type";
import { DoubleValueType } from "./double-value-type";



export class DoubleValueCapability implements Capability {
  id: string;
  minValue: number = 0;
  maxValue: number = 0;
  value: number = 0;
  _editValue: number = 0;
  capabilityType: CapabilityType;
  doubleValueType: DoubleValueType;
  material: Material | undefined;
  constructor(settingCapability: SettingDoubleValueCapability) {
    this.id = settingCapability.id;
    this.capabilityType = CapabilityType.DoubleValue;
    this.doubleValueType = settingCapability.doubleValueType;
    this.value = 0;
    this.minValue = settingCapability.minValue;
    this.maxValue = settingCapability.maxValue;
  }
  get editValue() : number {
    return this._editValue;
  }
  set editValue(val : number) {
    this._editValue = val;
    this.material?.wsService.doubleValueSet(this.material.unit.id, this.material.id, this.id, this._editValue);
  }
  beginEdit() {
    this._editValue = this.value;
  }
  setState(valueState: CapabilityState): void {
    let state: DoubleValueState = <DoubleValueState>valueState;
    this.value = state.value;
  }
}
